Fixes #42643 allow keys with colons in JSON config

SteveDunn opened the pull request:
Each provider has their own `Separator` field (defaults to `:`).  If users have colons in their keys, they can provide `` ` `` (or similar) as the optional parameter to the various methods.

Fixes #42643

SteveDunn committed:
42643 Work in Progress attempt at getting this to work. Note - hard-coded jsonfileprovider to backtick which'll probably break lots of tests

SteveDunn commented on src/libraries/Microsoft.Extensions.Configuration.Json/src/JsonConfigurationFileParser.cs:
```diff
@@ -15,6 +15,7 @@ internal sealed class JsonConfigurationFileParser
 
         private readonly Dictionary<string, string?> _data = new Dictionary<string, string?>(StringComparer.OrdinalIgnoreCase);
         private readonly Stack<string> _paths = new Stack<string>();
+        private string _keyDelimiter = "`";// ConfigurationPath.KeyDelimiter;
```
Of course, by changing `:` to `` ` ``, means we can't have `` ` `` in keys.
And obviously, this will need to be configurable, with a default defaulting to `:`.

SteveDunn replied:
My other idea was for each `IConfigurationProvider` to specify its delimiter, but that would be a breaking change for older runtimes that don't support default interface implementation:
```csharp
    public interface IConfigurationProvider
    {
        public string GetDelimiter() => ":";
        ...
```

SteveDunn committed:
Can now have colons in keys in json, but it still breaks other tests. Need the provider to specify their delimiter (to do next)

SteveDunn committed:
Got backtick separator working for json provider.
Now need to remove previous effort of passing separators everywhere

SteveDunn committed:
Fixed separator - now all json tests pass,although combinations are broken

SteveDunn committed:
Fixed all but 1 test

SteveDunn committed:
All tests pass, although `GetChildrenImplementation` is still a mess

SteveDunn committed:
Tidy up (somewhat) `GetChildrenImplementation` to cater for the fact that each provider can have a separate key

SteveDunn committed:
Tidy

SteveDunn committed:
Tidy

AraHaan commented on src/libraries/Microsoft.Extensions.Configuration.Json/ref/Microsoft.Extensions.Configuration.Json.cs:
```diff
@@ -8,12 +8,14 @@ namespace Microsoft.Extensions.Configuration
-        public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddJsonFile(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, string path, bool optional) { throw null; }
-        public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddJsonFile(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, string path, bool optional, bool reloadOnChange) { throw null; }
-        public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddJsonStream(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, System.IO.Stream stream) { throw null; }
+        public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddJsonFile(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, string path, string separator = ":") { throw null; }
+        public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddJsonFile(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, string path, bool optional, string separator = ":") { throw null; }
+
+        //todo: SD: need to add separate to rest of methods here.
+        public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddJsonFile(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, string path, bool optional, bool reloadOnChange, string separator = ":") { throw null; }
+        public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddJsonStream(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, System.IO.Stream stream, string separator = ":") { throw null; }
```
Woud need to create overload version without these, this would result in an ABI break if merged.

SteveDunn committed:
PR feedback - duplicate signatures rather than use default parameters

SteveDunn committed:
Tidy up and fix issue where default delimiter was used for sorting

tebeco commented:
see https://github.com/dotnet/runtime/issues/42643#issuecomment-1075992066 for using array instead of direct keys

you might get more flexible code for the original need

maryamariyan commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -690,6 +690,8 @@ private static Array BindArray(Array source, IConfiguration config, BinderOption
         [RequiresUnreferencedCode(PropertyTrimmingWarningMessage)]
         private static object? GetPropertyValue(PropertyInfo property, object instance, IConfiguration config, BinderOptions options)
         {
+            // List<string> separators = config.Select(p => (p as ConfigurationProvider)?.GetDelimiter() ?? ":").Distinct().ToList();
```
nit: remove dead code, found more on other files

maryamariyan commented on src/libraries/Microsoft.Extensions.Configuration.Json/ref/Microsoft.Extensions.Configuration.Json.cs:
```diff
@@ -9,11 +9,17 @@ namespace Microsoft.Extensions.Configuration
         public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddJsonFile(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, string path) { throw null; }
+        public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddJsonFile(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, string path, string separator) { throw null; }
         public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddJsonFile(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, string path, bool optional) { throw null; }
+        public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddJsonFile(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, string path, bool optional, string separator) { throw null; }
```
You shouldn't need to manually update the `\ref\` files.

Please use the instruction in [updating-ref-source.md#for-most-assemblies-within-libraries](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/docs/coding-guidelines/updating-ref-source.md#for-most-assemblies-within-libraries) instead.

If your fix needs new API addition it would first need to be proposed through [API Review Process](https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md), that said I'm not sure yet why we would need these new APIs added.

SteveDunn replied:
Have regenerated that file. I deleted the contents of it before running, just to ensure there were no remnants left in.

maryamariyan commented on src/libraries/Microsoft.Extensions.Configuration.Json/tests/JsonConfigurationTest.cs:
```diff
@@ -38,6 +39,29 @@ public void CanLoadValidJsonFromStreamProvider()
+            public string Uri { get; set; }
+        }
+
+
```
nit: extra blank line

maryamariyan commented on src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs:
```diff
@@ -34,7 +34,8 @@ public FileConfigurationProvider(FileConfigurationSource source!!)
                     {
                         Thread.Sleep(Source.ReloadDelay);
-                        Load(reload: true);
+                        throw new NotImplementedException("Need to know the separator string here...");
```
This seems out of place

SteveDunn replied:
Oops! Sorry, I'll look at that - must've overlooked it as I was tidying up

maryamariyan commented on src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.cs:
```diff
@@ -26,9 +26,11 @@ public static partial class ConfigurationPath
         public static readonly string KeyDelimiter;
         public static string Combine(System.Collections.Generic.IEnumerable<string> pathSegments) { throw null; }
         public static string Combine(params string[] pathSegments) { throw null; }
+        public static string CombineWith(string separator, params string[] pathSegments) { throw null; }
```
Why would you need to add new APIs here? I wasn't expecting new API added for this issue.

SteveDunn replied:
The new method is because the `ConfigurationSection` can now have separators that are not `:`.  The original `Combine` method assumed that it would always be `:`, which is no longer the case.  This is no longer the case as users can specify a different separator, e.g. `` ` ``, so that they can have key names with colons

SteveDunn replied:
Is there anything else needed for this PR? It fixes the issue reported and doesn't break any existing functionality. There's new API's, but they're needed if we want key separators to be configurable.

maryamariyan replied:
Would you be open to summarizing the result of this PR into the issue page Something with a structure similar to the description in https://github.com/dotnet/runtime/issues/15725? (containing Summary, Rationale, Usage, and Proposed API). Then let's close this PR.

SteveDunn replied:
@maryamariyan - sure, https://github.com/dotnet/runtime/issues/67616

Please let me know if it can be improved.

maryamariyan reviewed (changes requested):
Is this PR still work in progress? I switched it to a draft for now.

SteveDunn commented:
> Is this PR still work in progress? I switched it to a draft for now.

I moved it from WIP as all it fixes the problem reported in the Issue. As you can see though, it did require some overloads to allow the user to specify their new separator string for the keys.  I couldn't think of another way of allowing this, so I'm open to suggestions and ideas.

SteveDunn committed:
Removed code that should've been removed previously as part of the clean up

SteveDunn committed:
PR feedback: ran dotnet msbuild /t:GenerateReferenceAssemblySource to update the `ref` file

SteveDunn committed:
Remove commented code

SteveDunn committed:
Remove extra blank line

SteveDunn commented:
> > Is this PR still work in progress? I switched it to a draft for now.
> 
> I moved it from WIP as all it fixes the problem reported in the Issue. As you can see though, it did require some overloads to allow the user to specify their new separator string for the keys. I couldn't think of another way of allowing this, so I'm open to suggestions and ideas.

... and, 6 days later I find the button that really does move it from draft! 🤦

SteveDunn committed:
Merge branch 'main' into 42643-bind-to-json-containing-keys-with-colons

SteveDunn committed:
Post merge: delimiter is now per-instance in comparer rather than static

SteveDunn commented:
Closed for now until the API Proposal is reviewed:
https://github.com/dotnet/runtime/issues/67616
